File: lab3/laplsolv-parallel.f90
program laplsolv
    use omp_lib

    !-----------------------------------------------------------------------
    ! Serial program for solving the heat conduction problem 
    ! on a square using the Jacobi method. 
    ! Written by Fredrik Berntsson (frber@math.liu.se) March 2003
    ! Modified by Berkant Savas (besav@math.liu.se) April 2006
    ! Paralleled by Hao-Hsiang Liao May 2018
    ! Modified by Santi Pagola May 2018
    !-----------------------------------------------------------------------
    integer, parameter                        :: n=1000, maxiter=1000
    double precision,parameter                :: tol=1.0E-3
    double precision,dimension(0:n+1,0:n+1)   :: T
    double precision                          :: error
    double precision                          :: t1,t0
    integer                                   :: i,j,k
    integer                                   :: nr_threads
    integer                                   :: ratio,my_id
    integer,dimension(0:99)                   :: start_pos,stop_pos ! Initialize with a big number
    character(len=20)                         :: str
    double precision,dimension(n)             :: tmp2
    double precision,dimension(0:n-1,0:99)    :: tmp1,tmp3

    ! Set boundary conditions and initial values for the unknowns
    T=0.0D0
    T(0:n+1 , 0)     = 1.0D0
    T(0:n+1 , n+1)   = 1.0D0
    T(n+1   , 0:n+1) = 2.0D0

    !$omp parallel shared(nr_threads)
    nr_threads    = OMP_GET_NUM_THREADS()
    !$omp end parallel


    ! Determine how many stuffs should be implemented by a thread
    ratio = n/nr_threads

      ! Determine every thread's start/stop position
     do i=0,nr_threads-1
        start_pos(i) = (i*ratio)+1
        stop_pos(i) = (i+1)*ratio
     end do

  ! Solve the linear system of equations using the Jacobi method
  t0 = omp_get_wtime()
  
  do k=1,maxiter
     
     !tmp1=T(1:n,0)
     error=0.0D0

     ! We need to assign various start_pos/stop_pos's previous one array for each thread, their relationship is tmp1=T(1:n,j-1) when tmp2=T(1:n,j)
     ! tmp3 is an array that consist original stop_pos's value
     do i=0,nr_threads-1
        tmp1(0:n-1, i) = T(1:n, start_pos(i)-1)
        tmp3(0:n-1, i) = T(1:n, stop_pos(i)+1)
	 end do

     !$omp parallel private(j,tmp2,my_id) shared(T,tmp1,tmp3,k) reduction(max:error)
     my_id = OMP_GET_THREAD_NUM()

     do j=start_pos(my_id),(stop_pos(my_id)-1)
        tmp2=T(1:n,j)
        T(1:n,j)=(T(0:n-1,j)+T(2:n+1,j)+T(1:n,j+1)+tmp1(0:n-1,my_id))/4.0D0
        error=max(error,maxval(abs(tmp2-T(1:n,j))))
        tmp1(0:n-1,my_id)=tmp2
     end do
     
     tmp2=T(1:n,stop_pos(my_id))
     T(1:n,j)=(T(0:n-1,j)+T(2:n+1,j)+tmp3(0:n-1,my_id)+tmp1(0:n-1,my_id))/4.0D0
     error=max(error,maxval(abs(tmp2-T(1:n,j))))
     tmp1(0:n-1,my_id)=tmp2
     
     !$omp end parallel

     if (error<tol) then
        exit
     end if
     
  end do
  
  t1 = omp_get_wtime()

  print *, 'The number of threads is',nr_threads
  write(unit=*,fmt=*) 'Time:',t1-t0,'Number of Iterations:',k
  write(unit=*,fmt=*) 'Temperature of element T(1,1)  =',T(1,1)

  ! Uncomment the next part if you want to write the whole solution
  ! to a file. Useful for plotting. 
  
  !open(unit=7,action='write',file='result.dat',status='unknown')
  !write(unit=str,fmt='(a,i6,a)') '(',N,'F10.6)'
  !do i=0,n+1
  !   write (unit=7,fmt=str) T(i,0:n+1)  
  !end do
  !close(unit=7)
  
end program laplsolv

File: lab3/laplsolv-parallel.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#include <omp.h>

double find_max(double * array, size_t n)
{
    double max = array[0];
    for (size_t i = 1; i < n; ++i)
    {
        if (array[i] > max)
            max = array[i];
    }
    return max;
}

void init_big_T(double ** T, int rows, int cols)
{
    for (int i = 0; i < rows; ++i)
    {
        for (int j = 0; j < cols; ++j)
        {
            // Always, j=0 or j==n-1 means a 1 must be set
            if (j == 0 || j == cols - 1)
            {
                if (i < rows - 1) 
                    T[i][j] = 1.0;
                else
                    T[i][j] = 2.0;
            }
            else
            {
                if (i == rows - 1)
                    T[i][j] = 2.0;
                else
                    T[i][j] = 0.0;
            }
        }
    }
}

int main(int argc, char* argv[])
{
    int nr_threads;
    int n = 1000;
    int j;
    int my_id;
    int maxiter = 1000;
    double tol = 1.0e-3;
    /* Allocate a big matrix */
    double ** T = (double **) malloc (sizeof *T * (n + 2));
    for (int i = 0 ; i < n + 2; ++i)
        *(T + i) = (double *) malloc (sizeof(double) * (n + 2));
    double error, t0, t1;
    int * start_pos;
    int * end_pos;
    double *tmp2;
#pragma omp parallel
    {
        nr_threads = omp_get_num_threads();

        /* Allocate tmp2 - array */
        tmp2 = (double *) malloc (sizeof *tmp2 * n);
    }
    /* Allocate tmp's 1 and 3 - matrices */
    double ** tmp1 = (double ** ) malloc (sizeof *tmp1 * n);
    double ** tmp3 = (double ** ) malloc (sizeof *tmp3 * n);
    for (int i = 0; i < n; ++i)
        *(tmp1 + i) = (double *) malloc (sizeof (double) * nr_threads);
    for (int i = 0; i < n; ++i)
        *(tmp3 + i) = (double *) malloc (sizeof (double) * nr_threads);

    /* Set boundary conditions and initial values for the unknowns */
    init_big_T(T, n + 2, n + 2);

    int ratio = n / nr_threads;

    // Init the start & end positions
    start_pos = (int * ) malloc (sizeof *start_pos * nr_threads);
    end_pos   = (int * ) malloc (sizeof *end_pos * nr_threads);

    for (int i = 0; i < nr_threads; ++i)
    {
        start_pos[i] = (i * ratio) + 1;
        end_pos[i]   = (i + 1) * ratio;
    }

    t0 = omp_get_wtime();

    int iter; 
    for (iter = 0; iter <= maxiter; ++iter)
    {
        error = 0.0;

        for (int i = 0; i < nr_threads; ++i) // cols in big T
        {
            for (int k = 1; k <= n; ++k) // rows in big T
            {
                tmp1[k-1][i] = T[k][start_pos[i] - 1];
                tmp3[k-1][i] = T[k][end_pos[i] + 1];
            }
        }
#pragma omp parallel private(j,my_id) shared(T,tmp1,tmp3,tmp2)
        {
            my_id = omp_get_thread_num();
            double * diff = (double *) malloc(sizeof *diff * n);
            double current_error = 0.0;
            for ( int pos = start_pos[my_id]; pos < end_pos[my_id]; ++pos )
            {
                for (int k = 0; k < n; ++k)
                {
                    tmp2[k] = T[k+1][pos];
                    T[k+1][pos] = (T[k][pos] + T[k+2][pos] + T[k+1][pos+1] + tmp1[k][my_id]) / 4.0;
                    diff[k]  = fabs(tmp2[k] - T[k+1][pos]);
                    tmp1[k][my_id] = tmp2[k];
                }
                current_error = find_max(diff, n);
                /* Get the maximum */
                error = error > current_error ? error : current_error;
            }
            for ( int k = 0; k < n; ++k)
            {
                tmp2[k] = T[k+1][end_pos[my_id]];
                T[k+1][end_pos[my_id]] =
                    (T[k][end_pos[my_id]] + T[k+2][end_pos[my_id]] + tmp3[k][my_id] + tmp1[k][my_id]) / 4.0;
                diff[k] = fabs(tmp2[k] - T[k+1][end_pos[my_id]]);
                tmp1[k][my_id] = tmp2[k];
            }
            current_error = find_max(diff, n);
            error = error > current_error ? error : current_error;

            free(diff);
        }
        if (error < tol)
            break;
    }

    t1 = omp_get_wtime();

    printf("-----------------------\n");
    printf("Elapsed:\t%.6f s.\n", t1 - t0);
    printf("Iterations:\t%d\n", iter);
    printf("Temperature:\t%.12f\n", T[1][1]);
    printf("Nr. threads:\t%d\n", nr_threads);
    printf("-----------------------\n");

    /* Free stuff */
    for (int i = 0; i < n + 2; ++i)
        free(*(T + i));
    free(T);

    return 0;
}

